In [111]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, OneHotEncoder

In [112]:
df=pd.read_excel('loan status.csv.xlsx')
df

Loan_ID  Gender Married Dependents     Education Self_Employed  \
0    LP001002    Male      No          0      Graduate            No   
1    LP001003    Male     Yes          1      Graduate            No   
2    LP001005    Male     Yes          0      Graduate           Yes   
3    LP001006    Male     Yes          0  Not Graduate            No   
4    LP001008    Male      No          0      Graduate            No   
..        ...     ...     ...        ...           ...           ...   
609  LP002978  Female      No          0      Graduate            No   
610  LP002979    Male     Yes         3+      Graduate            No   
611  LP002983    Male     Yes          1      Graduate            No   
612  LP002984    Male     Yes          2      Graduate            No   
613  LP002990  Female      No          0      Graduate           Yes   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0               5849                0.0         NaN             360.0   
1               4583             1508.0       128.0             360.0   
2               3000                0.0        66.0             360.0   
3               2583             2358.0       120.0             360.0   
4               6000                0.0       141.0             360.0   
..               ...                ...         ...               ...   
609             2900                0.0        71.0             360.0   
610             4106                0.0        40.0             180.0   
611             8072              240.0       253.0             360.0   
612             7583                0.0       187.0             360.0   
613             4583                0.0       133.0             360.0   

     Credit_History Property_Area Loan_Status  
0               1.0         Urban           Y  
1               1.0         Rural           N  
2               1.0         Urban           Y  
3               1.0         Urban           Y  
4               1.0         Urban           Y  
..              ...           ...         ...  
609             1.0         Rural           Y  
610             1.0         Rural           Y  
611             1.0         Urban           Y  
612             1.0         Urban           Y  
613             0.0     Semiurban           N  

[614 rows x 13 columns]

In [113]:
df.columns

Index(['Loan_ID', 'Gender', 'Married', 'Dependents', 'Education',
       'Self_Employed', 'ApplicantIncome', 'CoapplicantIncome', 'LoanAmount',
       'Loan_Amount_Term', 'Credit_History', 'Property_Area', 'Loan_Status'],
      dtype='object')

In [114]:
X=df[['Gender', 'Married', 'Dependents', 'Education',
       'Self_Employed', 'ApplicantIncome', 'CoapplicantIncome', 'LoanAmount',
       'Loan_Amount_Term', 'Credit_History', 'Property_Area']]


In [115]:
y=df.Loan_Status

In [116]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

In [117]:
X_train.shape

(411, 11)

In [118]:
y_test.shape

(203,)

### Preprocessing train set

In [119]:
X_train.isnull().sum()

Gender               11
Married               1
Dependents           11
Education             0
Self_Employed        17
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount           15
Loan_Amount_Term     11
Credit_History       31
Property_Area         0
dtype: int64

In [120]:
# filling all nan values of categorical variables

gmode=X_train['Gender'].mode().iloc[0]
dmode=X_train['Dependents'].mode().iloc[0]
mmode=X_train['Married'].mode().iloc[0]
smode=X_train['Self_Employed'].mode().iloc[0]
cmode=X_train['Credit_History'].mode().iloc[0]
lamode=X_train['Loan_Amount_Term'].mode().iloc[0]


X_train['Gender'].fillna(gmode,inplace=True)
X_train['Dependents'].fillna(dmode,inplace=True)
X_train['Married'].fillna(mmode,inplace=True)
X_train['Self_Employed'].fillna(smode,inplace=True)
X_train['Credit_History'].fillna(cmode,inplace=True)
X_train['Loan_Amount_Term'].fillna(lamode,inplace=True)

D:\Anaconda3\lib\site-packages\pandas\core\generic.py:6245: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._update_inplace(new_data)


In [121]:
# filling nan values of numerical variables

X_train['LoanAmount'].fillna(X_train['LoanAmount'].median(),inplace=True)

In [122]:
X_train.isnull().sum()

Gender               0
Married              0
Dependents           0
Education            0
Self_Employed        0
ApplicantIncome      0
CoapplicantIncome    0
LoanAmount           0
Loan_Amount_Term     0
Credit_History       0
Property_Area        0
dtype: int64

In [123]:
# Transforming the values of dependents from 3+ to 3

X_train.loc[X_train['Dependents']=='3+','Dependents']=3
X_train['Dependents'].value_counts()

D:\Anaconda3\lib\site-packages\pandas\core\indexing.py:966: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s


0    246
1     67
2     63
3     35
Name: Dependents, dtype: int64

### Preprocessing test set

In [124]:
X_test

Gender Married Dependents     Education Self_Employed  ApplicantIncome  \
350    Male     Yes          0      Graduate            No             9083   
377    Male     Yes          0      Graduate            No             4310   
163    Male     Yes          2      Graduate            No             4167   
609  Female      No          0      Graduate            No             2900   
132    Male      No          0      Graduate            No             2718   
..      ...     ...        ...           ...           ...              ...   
280    Male      No          1  Not Graduate           Yes             4053   
195    Male     Yes          1      Graduate            No             3125   
613  Female      No          0      Graduate           Yes             4583   
495  Female     Yes          1      Graduate            No             3326   
451    Male     Yes          2  Not Graduate            No             1958   

     CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
350                0.0       228.0             360.0             1.0   
377                0.0       130.0             360.0             NaN   
163             1447.0       158.0             360.0             1.0   
609                0.0        71.0             360.0             1.0   
132                0.0        70.0             360.0             1.0   
..                 ...         ...               ...             ...   
280             2426.0       158.0             360.0             0.0   
195             2583.0       170.0             360.0             1.0   
613                0.0       133.0             360.0             0.0   
495              913.0       105.0              84.0             1.0   
451             1456.0        60.0             300.0             NaN   

    Property_Area  
350     Semiurban  
377     Semiurban  
163         Rural  
609         Rural  
132     Semiurban  
..            ...  
280         Urban  
195     Semiurban  
613     Semiurban  
495     Semiurban  
451         Urban  

[203 rows x 11 columns]

In [125]:
y_test

350    Y
377    Y
163    Y
609    Y
132    Y
      ..
280    N
195    N
613    N
495    Y
451    Y
Name: Loan_Status, Length: 203, dtype: object

In [126]:
X_test.isnull().sum()

Gender                2
Married               2
Dependents            4
Education             0
Self_Employed        15
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount            7
Loan_Amount_Term      3
Credit_History       19
Property_Area         0
dtype: int64

In [127]:
# filling all nan values of categorical variables

gmode=X_test['Gender'].mode().iloc[0]
dmode=X_test['Dependents'].mode().iloc[0]
mmode=X_test['Married'].mode().iloc[0]
smode=X_test['Self_Employed'].mode().iloc[0]
cmode=X_test['Credit_History'].mode().iloc[0]
lamode=X_test['Loan_Amount_Term'].mode().iloc[0]


X_test['Gender'].fillna(gmode,inplace=True)
X_test['Dependents'].fillna(dmode,inplace=True)
X_test['Married'].fillna(mmode,inplace=True)
X_test['Self_Employed'].fillna(smode,inplace=True)
X_test['Credit_History'].fillna(cmode,inplace=True)
X_test['Loan_Amount_Term'].fillna(lamode,inplace=True)

D:\Anaconda3\lib\site-packages\pandas\core\generic.py:6245: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._update_inplace(new_data)


In [128]:
# filling nan values of numerical variables

X_test['LoanAmount'].fillna(X_test['LoanAmount'].median(),inplace=True)

In [129]:
X_test.isnull().sum()

Gender               0
Married              0
Dependents           0
Education            0
Self_Employed        0
ApplicantIncome      0
CoapplicantIncome    0
LoanAmount           0
Loan_Amount_Term     0
Credit_History       0
Property_Area        0
dtype: int64

In [130]:
# Transforming the values of dependents from 3+ to 3

X_test.loc[X_test['Dependents']=='3+','Dependents']=3
X_test['Dependents'].value_counts()

D:\Anaconda3\lib\site-packages\pandas\core\indexing.py:966: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s


0    114
2     38
1     35
3     16
Name: Dependents, dtype: int64

## Encoding

### label encoding X_train 


In [131]:
# label encoding X_train 

lgetr=LabelEncoder()
lmatr=LabelEncoder()
ledtr=LabelEncoder()
lsetr=LabelEncoder()


X_train.Gender=lgetr.fit_transform(X_train['Gender'])
X_train.Married=lmatr.fit_transform(X_train['Married'])
X_train.Education=ledtr.fit_transform(X_train['Education'])
X_train.Self_Employed=lsetr.fit_transform(X_train['Self_Employed'])

D:\Anaconda3\lib\site-packages\pandas\core\generic.py:5303: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [132]:
X_train

Gender  Married Dependents  Education  Self_Employed  ApplicantIncome  \
244       1        1          0          1              0             3406   
393       1        1          2          1              0             1993   
310       0        0          0          0              0             2917   
408       1        1          1          0              0             8300   
572       1        1          2          0              0            16666   
..      ...      ...        ...        ...            ...              ...   
71        1        1          2          1              1             1875   
106       1        1          2          0              0            11417   
270       0        0          0          0              0             3237   
435       0        1          0          0              0            10047   
102       1        1          0          0              0            13650   

     CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
244             4417.0       123.0             360.0             1.0   
393             1625.0       113.0             180.0             1.0   
310                0.0        84.0             360.0             1.0   
408                0.0       152.0             300.0             0.0   
572                0.0       275.0             360.0             1.0   
..                 ...         ...               ...             ...   
71              1875.0        97.0             360.0             1.0   
106             1126.0       225.0             360.0             1.0   
270                0.0        30.0             360.0             1.0   
435                0.0       127.5             240.0             1.0   
102                0.0       127.5             360.0             1.0   

    Property_Area  
244     Semiurban  
393     Semiurban  
310     Semiurban  
408     Semiurban  
572         Urban  
..            ...  
71      Semiurban  
106         Urban  
270         Urban  
435     Semiurban  
102         Urban  

[411 rows x 11 columns]

In [133]:
# ordinal encoding X_train

la={12.0:0,36.0:1,60.0:2,84.0:3,120.0:4,180.0:5,240.0:6,300.0:7,360.0:8,480.0:9}
pa={'Rural':0,'Semiurban':1,'Urban':2}

X_train['Loan_Amount_Term']=X_train['Loan_Amount_Term'].map(la)
X_train['Property_Area']=X_train['Property_Area'].map(pa)

<ipython-input-133-c710ec724e8d>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X_train['Loan_Amount_Term']=X_train['Loan_Amount_Term'].map(la)
<ipython-input-133-c710ec724e8d>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X_train['Property_Area']=X_train['Property_Area'].map(pa)


In [134]:
X_train

Gender  Married Dependents  Education  Self_Employed  ApplicantIncome  \
244       1        1          0          1              0             3406   
393       1        1          2          1              0             1993   
310       0        0          0          0              0             2917   
408       1        1          1          0              0             8300   
572       1        1          2          0              0            16666   
..      ...      ...        ...        ...            ...              ...   
71        1        1          2          1              1             1875   
106       1        1          2          0              0            11417   
270       0        0          0          0              0             3237   
435       0        1          0          0              0            10047   
102       1        1          0          0              0            13650   

     CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
244             4417.0       123.0                 8             1.0   
393             1625.0       113.0                 5             1.0   
310                0.0        84.0                 8             1.0   
408                0.0       152.0                 7             0.0   
572                0.0       275.0                 8             1.0   
..                 ...         ...               ...             ...   
71              1875.0        97.0                 8             1.0   
106             1126.0       225.0                 8             1.0   
270                0.0        30.0                 8             1.0   
435                0.0       127.5                 6             1.0   
102                0.0       127.5                 8             1.0   

     Property_Area  
244              1  
393              1  
310              1  
408              1  
572              2  
..             ...  
71               1  
106              2  
270              2  
435              1  
102              2  

[411 rows x 11 columns]

### label encoding X_test


In [135]:
# label encoding X_test

lgetr=LabelEncoder()
lmatr=LabelEncoder()
ledtr=LabelEncoder()
lsetr=LabelEncoder()


X_test.Gender=lgetr.fit_transform(X_test['Gender'])
X_test.Married=lmatr.fit_transform(X_test['Married'])
X_test.Education=ledtr.fit_transform(X_test['Education'])
X_test.Self_Employed=lsetr.fit_transform(X_test['Self_Employed'])

D:\Anaconda3\lib\site-packages\pandas\core\generic.py:5303: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [136]:
X_test

Gender  Married Dependents  Education  Self_Employed  ApplicantIncome  \
350       1        1          0          0              0             9083   
377       1        1          0          0              0             4310   
163       1        1          2          0              0             4167   
609       0        0          0          0              0             2900   
132       1        0          0          0              0             2718   
..      ...      ...        ...        ...            ...              ...   
280       1        0          1          1              1             4053   
195       1        1          1          0              0             3125   
613       0        0          0          0              1             4583   
495       0        1          1          0              0             3326   
451       1        1          2          1              0             1958   

     CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
350                0.0       228.0             360.0             1.0   
377                0.0       130.0             360.0             1.0   
163             1447.0       158.0             360.0             1.0   
609                0.0        71.0             360.0             1.0   
132                0.0        70.0             360.0             1.0   
..                 ...         ...               ...             ...   
280             2426.0       158.0             360.0             0.0   
195             2583.0       170.0             360.0             1.0   
613                0.0       133.0             360.0             0.0   
495              913.0       105.0              84.0             1.0   
451             1456.0        60.0             300.0             1.0   

    Property_Area  
350     Semiurban  
377     Semiurban  
163         Rural  
609         Rural  
132     Semiurban  
..            ...  
280         Urban  
195     Semiurban  
613     Semiurban  
495     Semiurban  
451         Urban  

[203 rows x 11 columns]

In [137]:
# ordinal encoding X_test

la={12.0:0,36.0:1,60.0:2,84.0:3,120.0:4,180.0:5,240.0:6,300.0:7,360.0:8,480.0:9}
pa={'Rural':0,'Semiurban':1,'Urban':2}

X_test['Loan_Amount_Term']=X_test['Loan_Amount_Term'].map(la)
X_test['Property_Area']=X_test['Property_Area'].map(pa)

<ipython-input-137-9d49b0220516>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X_test['Loan_Amount_Term']=X_test['Loan_Amount_Term'].map(la)
<ipython-input-137-9d49b0220516>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X_test['Property_Area']=X_test['Property_Area'].map(pa)


In [138]:
X_test

Gender  Married Dependents  Education  Self_Employed  ApplicantIncome  \
350       1        1          0          0              0             9083   
377       1        1          0          0              0             4310   
163       1        1          2          0              0             4167   
609       0        0          0          0              0             2900   
132       1        0          0          0              0             2718   
..      ...      ...        ...        ...            ...              ...   
280       1        0          1          1              1             4053   
195       1        1          1          0              0             3125   
613       0        0          0          0              1             4583   
495       0        1          1          0              0             3326   
451       1        1          2          1              0             1958   

     CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
350                0.0       228.0                 8             1.0   
377                0.0       130.0                 8             1.0   
163             1447.0       158.0                 8             1.0   
609                0.0        71.0                 8             1.0   
132                0.0        70.0                 8             1.0   
..                 ...         ...               ...             ...   
280             2426.0       158.0                 8             0.0   
195             2583.0       170.0                 8             1.0   
613                0.0       133.0                 8             0.0   
495              913.0       105.0                 3             1.0   
451             1456.0        60.0                 7             1.0   

     Property_Area  
350              1  
377              1  
163              0  
609              0  
132              1  
..             ...  
280              2  
195              1  
613              1  
495              1  
451              2  

[203 rows x 11 columns]

### label encoding y_train 


In [139]:
y_train

244    Y
393    Y
310    Y
408    N
572    Y
      ..
71     Y
106    Y
270    Y
435    Y
102    Y
Name: Loan_Status, Length: 411, dtype: object

In [140]:
# label encoding y_train 

le=LabelEncoder()
y_train=le.fit_transform(y_train)

In [141]:
y_train

array([1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1,
       1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1,
       1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1,
       1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1,
       1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1,
       1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0,
       1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0,
       1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1,
       1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 0, 1, 1,
       1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1,

In [142]:
y_train=pd.Series(y_train)
y_train

0      1
1      1
2      1
3      0
4      1
      ..
406    1
407    1
408    1
409    1
410    1
Length: 411, dtype: int32

### label encoding y_test


In [143]:
y_test

350    Y
377    Y
163    Y
609    Y
132    Y
      ..
280    N
195    N
613    N
495    Y
451    Y
Name: Loan_Status, Length: 203, dtype: object

In [144]:
# label encoding y_test 

le=LabelEncoder()
y_test=le.fit_transform(y_test)

In [145]:
y_test

array([1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1,
       1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0,
       1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1,
       1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1,
       1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0,
       1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1,
       1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0,
       0, 0, 0, 1, 1])

In [146]:
y_test=pd.Series(y_test)
y_test

0      1
1      1
2      1
3      1
4      1
      ..
198    0
199    0
200    0
201    1
202    1
Length: 203, dtype: int32

### Scaling

In [147]:
from sklearn.preprocessing import MinMaxScaler

In [148]:
X_train.head()

Gender  Married Dependents  Education  Self_Employed  ApplicantIncome  \
244       1        1          0          1              0             3406   
393       1        1          2          1              0             1993   
310       0        0          0          0              0             2917   
408       1        1          1          0              0             8300   
572       1        1          2          0              0            16666   

     CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
244             4417.0       123.0                 8             1.0   
393             1625.0       113.0                 5             1.0   
310                0.0        84.0                 8             1.0   
408                0.0       152.0                 7             0.0   
572                0.0       275.0                 8             1.0   

     Property_Area  
244              1  
393              1  
310              1  
408              1  
572              2

In [149]:
# scaling X_train

scaler = MinMaxScaler()
X_train.iloc[:,5:6] = scaler.fit_transform(X_train.iloc[:,5:6])  
X_train.iloc[:,6:7] = scaler.fit_transform(X_train.iloc[:,6:7])  
X_train.iloc[:,7:8] = scaler.fit_transform(X_train.iloc[:,7:8])
X_train

D:\Anaconda3\lib\site-packages\pandas\core\indexing.py:966: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s
D:\Anaconda3\lib\site-packages\pandas\core\indexing.py:966: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s
D:\Anaconda3\lib\site-packages\pandas\core\indexing.py:966: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.

Gender  Married Dependents  Education  Self_Employed  ApplicantIncome  \
244       1        1          0          1              0         0.040272   
393       1        1          2          1              0         0.022795   
310       0        0          0          0              0         0.034224   
408       1        1          1          0              0         0.100804   
572       1        1          2          0              0         0.204280   
..      ...      ...        ...        ...            ...              ...   
71        1        1          2          1              1         0.021336   
106       1        1          2          0              0         0.139357   
270       0        0          0          0              0         0.038182   
435       0        1          0          0              0         0.122412   
102       1        1          0          0              0         0.166976   

     CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
244           0.130538    0.155198                 8             1.0   
393           0.048024    0.140556                 5             1.0   
310           0.000000    0.098097                 8             1.0   
408           0.000000    0.197657                 7             0.0   
572           0.000000    0.377745                 8             1.0   
..                 ...         ...               ...             ...   
71            0.055413    0.117130                 8             1.0   
106           0.033277    0.304539                 8             1.0   
270           0.000000    0.019034                 8             1.0   
435           0.000000    0.161786                 6             1.0   
102           0.000000    0.161786                 8             1.0   

     Property_Area  
244              1  
393              1  
310              1  
408              1  
572              2  
..             ...  
71               1  
106              2  
270              2  
435              1  
102              2  

[411 rows x 11 columns]

In [150]:
# scaling X_test

scaler = MinMaxScaler()
X_test.iloc[:,5:6] = scaler.fit_transform(X_test.iloc[:,5:6])  
X_test.iloc[:,6:7] = scaler.fit_transform(X_test.iloc[:,6:7])  
X_test.iloc[:,7:8] = scaler.fit_transform(X_test.iloc[:,7:8])
X_test

D:\Anaconda3\lib\site-packages\pandas\core\indexing.py:966: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s
D:\Anaconda3\lib\site-packages\pandas\core\indexing.py:966: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s
D:\Anaconda3\lib\site-packages\pandas\core\indexing.py:966: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.

Gender  Married Dependents  Education  Self_Employed  ApplicantIncome  \
350       1        1          0          0              0         0.140558   
377       1        1          0          0              0         0.064948   
163       1        1          2          0              0         0.062683   
609       0        0          0          0              0         0.042613   
132       1        0          0          0              0         0.039729   
..      ...      ...        ...        ...            ...              ...   
280       1        0          1          1              1         0.060877   
195       1        1          1          0              0         0.046177   
613       0        0          0          0              1         0.069273   
495       0        1          1          0              0         0.049361   
451       1        1          2          1              0         0.027690   

     CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
350           0.000000    0.370558                 8             1.0   
377           0.000000    0.204738                 8             1.0   
163           0.034728    0.252115                 8             1.0   
609           0.000000    0.104907                 8             1.0   
132           0.000000    0.103215                 8             1.0   
..                 ...         ...               ...             ...   
280           0.058224    0.252115                 8             0.0   
195           0.061992    0.272420                 8             1.0   
613           0.000000    0.209814                 8             0.0   
495           0.021912    0.162437                 3             1.0   
451           0.034944    0.086294                 7             1.0   

     Property_Area  
350              1  
377              1  
163              0  
609              0  
132              1  
..             ...  
280              2  
195              1  
613              1  
495              1  
451              2  

[203 rows x 11 columns]

In [151]:
y_train

0      1
1      1
2      1
3      0
4      1
      ..
406    1
407    1
408    1
409    1
410    1
Length: 411, dtype: int32

In [152]:
y_test

0      1
1      1
2      1
3      1
4      1
      ..
198    0
199    0
200    0
201    1
202    1
Length: 203, dtype: int32

## Applying ML Algo

### 1. Logistic Regression

In [153]:
from sklearn.linear_model import LogisticRegression

In [154]:
logmodel = LogisticRegression()
logmodel.fit(X_train,y_train)

LogisticRegression()

In [155]:
predictions = logmodel.predict(X_test)

Evaluation

In [156]:
from sklearn.metrics import classification_report
print(classification_report(y_test,predictions))

              precision    recall  f1-score   support

           0       0.94      0.46      0.62        72
           1       0.77      0.98      0.86       131

    accuracy                           0.80       203
   macro avg       0.86      0.72      0.74       203
weighted avg       0.83      0.80      0.78       203



In [157]:
print(confusion_matrix(y_test,predictions))

[[ 33  39]
 [  2 129]]


### 2. Decison Tree

In [158]:
from sklearn.tree import DecisionTreeClassifier

In [159]:
dtree = DecisionTreeClassifier()

In [160]:
dtree.fit(X_train,y_train)

DecisionTreeClassifier()

Evaluation

In [161]:
predictions = dtree.predict(X_test)

In [162]:
from sklearn.metrics import classification_report,confusion_matrix

In [163]:
print(classification_report(y_test,predictions))

              precision    recall  f1-score   support

           0       0.57      0.54      0.55        72
           1       0.75      0.77      0.76       131

    accuracy                           0.69       203
   macro avg       0.66      0.66      0.66       203
weighted avg       0.69      0.69      0.69       203



In [164]:
print(confusion_matrix(y_test,predictions))

[[ 39  33]
 [ 30 101]]


### 3. Random Forest

In [165]:
from sklearn.ensemble import RandomForestClassifier
rfc = RandomForestClassifier(n_estimators=100)
rfc.fit(X_train, y_train)

RandomForestClassifier()

In [166]:
rfc_pred = rfc.predict(X_test)

In [167]:
print(classification_report(y_test,rfc_pred))

              precision    recall  f1-score   support

           0       0.80      0.50      0.62        72
           1       0.77      0.93      0.84       131

    accuracy                           0.78       203
   macro avg       0.79      0.72      0.73       203
weighted avg       0.78      0.78      0.76       203



In [168]:
print(confusion_matrix(y_test,rfc_pred))

[[ 36  36]
 [  9 122]]


Accuracy

Logistic Regression :  80% <br>
Decision Tree :        67% <br>
Random Forest :        79% <br>

From the acuuracy of different models we can conclude that Logistic Regression is the best model that we can use for this dataset.